src/System.Linq.Expressions/src/System/Linq/Expressions/MemberAssignment.cs

```diff
@@ -81,28 +81,31 @@ public static MemberAssignment Bind(MethodInfo propertyAccessor, Expression expr
 
         private static void ValidateSettableFieldOrPropertyMember(MemberInfo member, out Type memberType)
         {
-            FieldInfo fi = member as FieldInfo;
-            if (fi == null)
+            Type decType = member.DeclaringType;
+            if (decType == null)
             {
-                PropertyInfo pi = member as PropertyInfo;
-                if (pi == null)
-                {
-                    throw Error.ArgumentMustBeFieldInfoOrPropertyInfo(nameof(member));
-                }
-                if (!pi.CanWrite)
-                {
-                    throw Error.PropertyDoesNotHaveSetter(pi, nameof(member));
-                }
-                memberType = pi.PropertyType;
+                throw Error.NotAMemberOfAnyType(member, nameof(member));
             }
-            else
+
+            // Null paramName as there are two paths here with different parameter names at the API
+            TypeUtils.ValidateType(decType, null);
+            switch (member)
             {
-                if (fi.DeclaringType == null)
-                {
-                    throw Error.NotAMemberOfAnyType(fi, nameof(member));
-                }
+                case PropertyInfo pi:
+                    if (!pi.CanWrite)
+                    {
+                        throw Error.PropertyDoesNotHaveGetter(pi, nameof(member));
```

Comment:
>PropertyDoesNotHaveGetter [](start = 36, length = 25)

Is this intended? PropertyDoesNotHaveSetter to PropertyDoesNotHaveGetter? can you please explain?